File: 1 course/1 semester/OAiP/C++/repos/LabFirstBlockThird/LabFirstBlockThird/LabFirstBlockThird.cpp
#include <iostream>
#include <iomanip>
#include <string>
#include <fstream>
#include <algorithm>

using namespace std;

int inputAmountOfOccurences() {
    int n;
    bool isIncorrect;
    const int MIN_NUM = 1;
    cout << "Введите число, количество вхождений строки 1 в строку 2, k = " << endl;
    do {
        isIncorrect = false;
        cin >> n;
        if (cin.fail())
        {
            cin.clear();
            while (cin.get() != '\n');
            cout << "Проверьте правильность ввода данных: " << endl;
            isIncorrect = true;
        }
        if (!isIncorrect && cin.get() != '\n')
        {
            cin.clear();
            while (cin.get() != '\n');
            cout << "Проверьте правильность ввода данных: " << endl;
            isIncorrect = true;
        }
        if (!isIncorrect && n < MIN_NUM) {
            cout << "Число > 0 " << endl;
            isIncorrect = true;
        }
    } while (isIncorrect);
    return n;
}

string chooseOption()
{
    bool isIncorrect;
    string input;
    cout << "Введите console, если хотите использовать консоль, file, если файл" << endl;
    do
    {
        isIncorrect = false;
        cin >> input;
        transform(input.begin(), input.end(), input.begin(), ::tolower);
        if (input != "console" && input != "file")
        {
            cerr << "Повторите ввод" << endl;
            isIncorrect = true;
        }
    } while (isIncorrect);
    return input;
}

string chooseConsoleWayToFill()
{
    bool isIncorrect;
    string input;
    cout << "Введите console, если хотите ввести данные сами, random, если хотите сгенерировать рандомные строки" << endl;
    do
    {
        isIncorrect = false;
        cin >> input;
        transform(input.begin(), input.end(), input.begin(), ::tolower);
        if (input != "console" && input != "random")
        {
            cerr << "Повторите ввод" << endl;
            isIncorrect = true;
        }
    } while (isIncorrect);
    return input;
}

string inputStrConsole()
{
    string str;
    bool isIncorrect;
    do
    {
        isIncorrect = false;
        cin >> str;
        if (str.empty() || str[0] == ' ')
        {
            isIncorrect = true;
            cerr << "Повторите ввод" << endl;
        }
    } while (isIncorrect);
    return str;
}

int calculateAnswer(string str1, string str2, int k)
{
    int countN = 0, n = 0;
    for (int i = 0; i <= str2.length() - str1.length(); i++)
    {
        string str = str2.substr(i, str1.length());
        if (str == str1) // (int j = str2.find(str1, i) == i)
        {
            ++countN;
            if (countN == k)
            {
                n = i + 1;
            }
        }
    }
    return n;
}

string randomStrConsole()
{
    int n;
    string str = "";
    string upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    string lower = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
    string numbers = "0123456789";
    string allCurrent = upper + lower + numbers;
    n = rand() % 256;
    for (int i = 0; i < n; i++)
    {
        str += allCurrent[rand() % 62];
    }
    return str;
}

int consoleInput(string input)
{
    bool isLengthRight;
    int k, n;
    string str1, str2;
    if (input == "console")
    {
        do {
            cerr << "Введите строку 1(первый символ не должен быть пробелом)" << endl;
            str1 = inputStrConsole();
            cout << "Введите строку 2(первый символ не должен быть пробелом)" << endl;
            str2 = inputStrConsole();
            isLengthRight = (str1.length() > str2.length()) ? true : false;
            if (isLengthRight)
            {
                cerr << "Повторите ввод строк, чтобы размер вашей первой строки был МЕНЬШЕ размера второй строки" << endl;
            }
        } while (isLengthRight);
    }
    else
    {
        do {
            str1 = randomStrConsole();
            str2 = randomStrConsole();
            isLengthRight = (str1.length() > str2.length()) ? true : false;
        } while (isLengthRight);
        cout << "Строки сгенерированы" << endl;
        cout << str1 << endl;
        cout << str2 << endl;
    }
    k = inputAmountOfOccurences();
    n = calculateAnswer(str1, str2, k);
    return n;
}

bool checkExtension(string path)
{
    bool isInCorrect = true;
    if ((path[path.length() - 1] == 't') && (path[path.length() - 2] == 'x') && (path[path.length() - 3] == 't') && (path[path.length() - 4] == '.'))
        isInCorrect = false;
    return isInCorrect;
}

bool isFileCorrect(string way)
{
    int k;
    bool isIncorrect = false;
    string str1, str2;
    ifstream fin(way);
    if (!fin.eof())
    {
        getline(fin, str1);
        if (str1.empty() || str1[0] == ' ')
        {
            isIncorrect = true;
            cerr << "Проверьте правильность первой строки" << endl;
        }
    }
    else
    {
        isIncorrect = true;
        cerr << "Нет данных для первой строки" << endl;
    }
    if (!fin.eof())
    {
        if (!isIncorrect)
        {
            getline(fin, str2);
            if (str2.empty() || str2[0] == ' ')
            {
                isIncorrect = true;
                cerr << "Проверьте правильность второй строки" << endl;
            }
        }
    }
    else
    {
        cerr << "Нет данных для второй строки" << endl;
        isIncorrect = true;
    }
    if (!fin.eof())
    {
        if (!isIncorrect)
        {
            fin >> k;
            if (fin.fail())
            {
                isIncorrect = true;
                cerr << "Число вхождений не является числом" << endl;
            }
            if (!isIncorrect && k < 1)
            {
                isIncorrect = true;
                cerr << "Число вхождений > 1" << endl;
            }
        }
    }
    if (!fin.eof() && !isIncorrect)
    {
        isIncorrect = true;
        cerr << "Уберите лишние данные" << endl;
    }
    fin.close();
    return isIncorrect;
}


string takeFileWay(string temp)
{
    string way;
    bool isIncorrect;
    do
    {
        isIncorrect = false;
        cout << "Введите путь к файлу" << endl;
        cin >> way;
        if (temp == "In")
        {
            ifstream fin(way);
            if (!fin.is_open() && checkExtension)
            {
                cout << "Проверьте правильность и нахождение вашего файла по заданному пути" << endl;
                isIncorrect = true;
            }
            else
            {
                isIncorrect = isFileCorrect(way);
            }
            fin.close();
        }
        else
        {
            ifstream fout(way);
            if (!fout.is_open() && checkExtension)
            {
                cout << "Проверьте правильность и нахождение вашего файла по заданному пути" << endl;
                    isIncorrect = true;
            }
            fout.close();
        }
    } while (isIncorrect);
    return way;
}

int fileInput(string way)
{
    int n, k;
    string str1, str2;
    ifstream fin(way, ios::beg);
    getline(fin, str1);
    getline(fin, str2);
    fin >> k;
    fin.close();
    n = calculateAnswer(str1, str2, k);
    return n;
}

int takeFinalInformation(string input)
{
    string way, str1, str2;
    int n;
    if (input == "console")
    {
        input = chooseConsoleWayToFill();
        n = consoleInput(input);
    }
    else
    {
        way = takeFileWay("In");
        n = fileInput(way);
    }
    return n;
}

void outputFinalInformation(int n, string outputChoice)
{
    string way;
    if (outputChoice == "console")
    {
        cout << "Программа завершилась успешно\n" << "номер позиции k-го вхождения: " << n;
    }
    else
    {
        way = takeFileWay("Out");
        ofstream fout(way);
        fout << "Программа завершилась успешно\n" << "номер позиции k-го вхождения: " << n;
        fout.close();
    }
}

int main()
{
    setlocale(LC_ALL, "RUSSIAN");
    int n;
    string inputChoice, outputChoice;
    cout << "Данная программа находит количество вхождений k 1-ой строки во 2-ую строку" << endl;
    inputChoice = chooseOption();
    n = takeFinalInformation(inputChoice);
    outputChoice = chooseOption();
    outputFinalInformation(n, outputChoice);
}

File: 1 course/1 semester/OAiP/Java/LabFirstBlockThird/src/LabFirstBlockThird.java
import java.util.Scanner;
import java.io.*;

public class LabFirstBlockThird {
    static Scanner scan = new Scanner(System.in);

    public static int inputAmountOfOccurences() {
        int n = 0;
        boolean isIncorrect;
        final int MIN_NUM = 1, MAX_NUM = 255;
        System.out.println("Введите число, количество вхождений строки 1 в строку 2, k = ");
        do {
            isIncorrect = false;
            try {
                n = Integer.parseInt(scan.nextLine());
            } catch (NumberFormatException E) {
                System.err.println("Повторите ввод: ");
                isIncorrect = true;
            }
            if (!isIncorrect && (n < MIN_NUM || n > MAX_NUM)) {
                System.err.println("Повторите ввод: ");
                isIncorrect = true;
            }
        } while (isIncorrect);
        return n;
    }

    public static String chooseOption(){
        boolean isIncorrect;
        String input;
        System.out.println("Введите console, если хотите использовать консоль, file, если файл");
        do {
            isIncorrect = false;
            input = scan.nextLine();
            if (!input.equalsIgnoreCase("console") && !input.equalsIgnoreCase("file")){
                System.err.println("Повторите ввод: ");
                isIncorrect = true;
            }
        } while (isIncorrect);
        return input;
    }

    public static String chooseConsoleWayToFill() {
        boolean isIncorrect;
        String input;
        System.out.println("Введите console, если хотите ввести данные с консоли, random, чтобы сгенерировать строки файл");
        do {
            isIncorrect = false;
            input = scan.nextLine();
            if (!input.equalsIgnoreCase("console") && !input.equalsIgnoreCase("random")){
                System.err.println("Повторите ввод: ");
                isIncorrect = true;
            }
        } while (isIncorrect);
        return input;
    }

    public static String inputStrConsole() {
        String str;
        boolean isIncorrect;
        do {
            isIncorrect = false;
            str = scan.nextLine();
            if (str.isEmpty() || str.charAt(0) == ' ') { //str.equals("");
                isIncorrect = true;
                System.err.println("Повторите ввод");
            }
        } while (isIncorrect);
        return str;
    }

    public static int calculateAnswer(String str1, String str2, int k) {
        int countN = 0, n = 0;
        for (int i = 0; i <= str2.length() - str1.length(); i++) {
            if (str2.substring(i, str1.length() + i).equals(str1)) {
                ++countN;
                if (countN == k) {
                    n = i + 1;
                }
            }
        }
        return n;
    }

    public static String randomStrConsole() {
        double n;
        String str = "", upper, lower, numbers, allCurrent;
        upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        lower = upper.toLowerCase();
        numbers = "0123456789";
        allCurrent = upper + lower + numbers;
        n = Math.random() * 255;
        for (int i = 0; i < n; i++) {
            str += allCurrent.charAt((int) Math.round(Math.random() * 61));
        }
        return str;
    }

    public static int consoleInput(String input) {
        boolean isLengthRight;
        int k, n;
        String str1 = "", str2 = "";
        if (input.equalsIgnoreCase("console")) {
            do {
                isLengthRight = false;
                System.out.println("Введите строку 1(первый символ не должен быть пробелом)");
                str1 = inputStrConsole();
                System.out.println("Введите строку 2(первый символ не должен быть пробелом)");
                str2 = inputStrConsole();
                if (str1.length() > str2.length()) {
                    System.err.println("Повторите ввод строк, чтобы размер вашей первой строки был МЕНЬШЕ размера второй строки");
                    isLengthRight = true;
                }
            } while (isLengthRight);
        } else {
            do {
                isLengthRight = false;
                str1 = randomStrConsole();
                str2 = randomStrConsole();
                if (str1.length() > str2.length()) {
                    isLengthRight = true;
                }
            } while (isLengthRight);
            System.out.println("Строки сгенерированы.");
            System.out.println(str1);
            System.out.println(str2);
        }
        k = inputAmountOfOccurences();
        n = calculateAnswer(str1, str2, k);
        return n;
    }

    public static boolean isTakeInformationFromFile(String way) {
        int k = 0;
        final int MIN_NUM = 1, MAX_NUM = 255;
        String str1, str2;
        boolean isIncorrect = false;
        try (BufferedReader br = new BufferedReader(new FileReader(way))) {
            if ((str1 = br.readLine()) != null) {
                if (str1.isEmpty() || str1.charAt(0) == ' ') { //str1.isEmpty()
                    isIncorrect = true;
                    System.err.println("Проверьте правильность введенной 1-ой строки");
                }
            } else {
                isIncorrect = true;
                System.err.println("В файле нет данных для 1-ой строки");
            }
            if (!isIncorrect & (str2 = br.readLine()) != null) {
                if (str2.isEmpty() || str2.charAt(0) == ' ') { //str2.isEmpty()
                    isIncorrect = true;
                    System.err.println("Проверьте правильность введенной 2-ой строки");
                }
            } else {
                isIncorrect = true;
                System.err.println("В файле нет данных для 2-ой строки");
            }
            if (!isIncorrect & str1.length() > str2.length()) {
                System.err.println("Размер вашей строки 2 должен быть БОЛЬШЕ размера строки 1");
                isIncorrect = true;
            }
            if (!isIncorrect) {
                try {
                    k = Integer.parseInt(br.readLine());
                } catch (NumberFormatException E) {
                    isIncorrect = true;
                    System.err.println("Проверьте соответствие k к целому числу");
                }
                if (!isIncorrect & (k < MIN_NUM || k > MAX_NUM)) {
                    isIncorrect = true;
                    System.err.println("Ваше число должно быть в диапазоне " + MIN_NUM + " до " + MAX_NUM);
                }
            }
            if (!isIncorrect & br.readLine() != null) {
                isIncorrect = true;
                System.out.println("Уберите лишние данные");
            }
        } catch (IOException e) {
            System.err.println("Ошибка файла");
            System.err.println(e.getMessage());
            isIncorrect = true;
        }
        return isIncorrect;
    }

    public static String takeFileWay(String temp) {
        String way;
        boolean isIncorrect;
        do {
            isIncorrect = false;
            System.out.println("Введите путь к файлу");
            way = scan.nextLine();
            File file = new File(way);
            if (!file.exists() || !file.canWrite() ||!file.canRead() || !way.endsWith(".txt") || file.isDirectory()) {
                isIncorrect = true;
                System.err.println("Файл не найден, или неверный формат файла; повторите попытку ввода");
            } else if (temp.equals("In")) {
                isIncorrect = isTakeInformationFromFile(way);
            }
        } while (isIncorrect);
        return way;
    }

    public static String inputStringFromFile(String way, int p) {
    String str = "";
        try (BufferedReader br = new BufferedReader(new FileReader(way))) {
            if (p == 1) {
                br.readLine();
            }
            str = br.readLine();
        } catch (IOException e) {
            System.err.println("Ошибка файла");
            System.err.println(e.getMessage());
        }
        return str;
    }

    public static int inputAmountOfOccurencesFromFile(String way) {
        int k = 0;
        try (BufferedReader br = new BufferedReader(new FileReader(way))) {
        br.readLine();
        br.readLine();
        k = Integer.parseInt(br.readLine());
        } catch (IOException e) {
            System.err.println("Ошибка файла");
            System.err.println(e.getMessage());
        }
        return k;
    }
        public static int takeFinalInformation(String input) {
        String way, str1, str2;
        int n = 0, k;
        if (input.equalsIgnoreCase("console")) {
            input = chooseConsoleWayToFill();
            n = consoleInput(input);
        } else {
            way = takeFileWay("In");
            str1 = inputStringFromFile(way, 0);
            str2 = inputStringFromFile(way, 1);
            k = inputAmountOfOccurencesFromFile(way);
            n = calculateAnswer(str1, str2, k);
            System.out.println("Данные считаны успешно");
        }
        return n;
    }

    public static void writeAnswerConsole(int n) {
        System.out.println("Программа завершилась успешно.\n" + "номер позиции k-го вхождения: " + n);
    }

    public static void writeAnswerFile(String way, int n) {
        try (PrintWriter pw = new PrintWriter(new FileWriter(way, false))) {
            pw.write("Программа завершилась успешно.\n" + "номер позиции k-го вхождения: " + n);
        } catch (IOException e) {
            System.err.println("Ошибка файла");
            System.err.println(e.getMessage());
        }
    }
    public static void outputAnswer(String outputChoice, int n) {
        String way;
        if (outputChoice.equalsIgnoreCase("console")) {
            writeAnswerConsole(n);
        } else {
            way = takeFileWay("Out");
            writeAnswerFile(way, n);
        }
    }

    public static void main(String[] args) {
    String inputChoice, outputChoice;
    int n;
    System.out.println("Данная программа находит количество вхождений(задаваемого пользователем) первой строки во вторую");
    inputChoice = chooseOption();
    n = takeFinalInformation(inputChoice);
    outputChoice = chooseOption();
    outputAnswer(outputChoice, n);
    }
}
